feat(artwork lists): contextual action menu + delete action

anandaroop opened the pull request:
The type of this PR is: **Feat**

This PR solves [FX-4553]

This adds the ability to delete an Artwork List.

## Contextual Menu

In order to get there I needed a new kind of [contextual menu](https://www.nngroup.com/articles/contextual-menus/) that is triggered from a ⋯ icon. 

I extracted this into what is now the first commit. (If desired I can further extract this new component into a separate PR.)

I was going for a simple API that can be used like this:

```jsx
const SomeContextualMenu = () => (
  <ContextualMenu>
    <ContextualMenuItem onClick={doSomething}>Item one</ContextualMenuItem>
    <ContextualMenuDivider />
    <ContextualMenuItem onClick={doSomething}>Item two</ContextualMenuItem>
    <ContextualMenuDivider />
    <ContextualMenuItem onClick={doSomething}>Item three</ContextualMenuItem>
  </ContextualMenu>
);
```

This now lives in `src/Components/ContextualMenu`, is completely generic, and is documented in the Force storybook:

<img width="100%" alt="story" src="https://user-images.githubusercontent.com/140521/219902626-fcce76b3-ee15-4c5e-aed5-fd666ba40159.png">

That can then be wired up with the the new Delete modal as follows.

## Desktop

https://user-images.githubusercontent.com/140521/219853242-498cd312-9c39-42d2-802c-36e2a169593a.mov

## mWeb

https://user-images.githubusercontent.com/140521/219853245-234e78ae-b855-4cd5-99cb-872d428cee88.mov

[FX-4553]: https://artsyproduct.atlassian.net/browse/FX-4553?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

damassi commented on src/Apps/CollectorProfile/Routes/Saves2/Components/CollectionActionsMenu.tsx:
```diff
@@ -0,0 +1,52 @@
+import {
+  MoreIcon,
```
We should grab icons from `@artsy/icons`

anandaroop commented on src/Apps/CollectorProfile/Routes/Saves2/Components/Actions/DeleteSavesModal.tsx:
```diff
@@ -0,0 +1,100 @@
+          </Button>
+        </Flex>
+      }
+      width={[
+        "100%",
+        753, // vs. 713px for create and edit modals and 673px for manage-artworks modal -- should be consistent (700?) or maybe just "auto"?
+      ]}
```
I think many of these modals will need a consistency pass at some point soon.

dimatretyak replied:
yep, agree with you 🤔

anandaroop commented on src/Components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -0,0 +1,59 @@
+      {({ anchorRef, anchorProps }) => {
+        return (
+          <Clickable
+            // @ts-expect-error TODO: fix this ref type error
+            ref={anchorRef}
```
Got stumped by this one.

anandaroop commented on src/Components/ContextualMenu/ContextualMenuItem.tsx:
```diff
@@ -0,0 +1,37 @@
+})`
+  user-select: none;
+
+  transition: background-color 150ms;
+  &:hover {
+    background-color: ${themeGet("colors.black5")};
+  }
```
🧐 🧐 🧐 not sure where to find a standard for this.

<img width="400" alt="transitions" src="https://user-images.githubusercontent.com/140521/219855042-e68121ff-8d6b-4014-afe6-7d81b1f46027.png">

anandaroop replied:
The styling of the contextual menu  could in general benefit from some discussion — happy to follow up with Ryan and iterate in another PR.

damassi replied:
Good call. Generally speaking it should be 250. Can we move this to a `constants.ts` file in palette and export? Might be worth updating all of these in palette, too: 

<img width="742" alt="Screenshot 2023-02-22 at 2 51 19 PM" src="https://user-images.githubusercontent.com/236943/220778595-9b94b42f-d1a6-4bd0-b2cf-99131d55f5d6.png">

damassi commented on src/Components/ContextualMenu/__stories__/ContextualMenu.story.tsx:
```diff
@@ -0,0 +1,158 @@
+      </ContextualMenuItem>
+    </ContextualMenu>
+  )
+}
```
Thanks for adding a story

dimatretyak commented on src/Apps/CollectorProfile/Routes/Saves2/Components/SavesArtworks.tsx:
```diff
@@ -98,6 +101,7 @@ export const SavesArtworksRefetchContainer = createRefetchContainer(
           page: { type: "Int", defaultValue: 1 }
           sort: { type: "CollectionArtworkSorts" }
         ) {
+        internalID
```
I think `internalID` is not used anywhere and we can safely delete it

dimatretyak commented on src/Components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -0,0 +1,59 @@
+            {...anchorProps}
+            px={2}
+            pt={1}
+            tras
```
I think you forgot to add the correct name for some prop 😅

dimatretyak commented on src/Components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -0,0 +1,59 @@
+            tras
+            aria-label="Open contextual menu"
+          >
+            <MoreIcon width={[24, 32]} height={[24, 32]} />
```
**suggestion**: create a reused variable for sizes. Something like this

```javascript
const MORE_ICON_SIZE = [24, 32]
...
<MoreIcon width={MORE_ICON_SIZE} height={MORE_ICON_SIZE} />
```

dimatretyak commented on src/Components/ContextualMenu/ContextualMenu.tsx:
```diff
@@ -0,0 +1,59 @@
+         * Each ContextualMenuItem must be able to close this
+         * parent ContextualMenu's Dropdown when it is clicked.
+         */
+        return React.Children.map(children, (child: ReactElement) => {
+          if (child.type === ContextualMenuItem) {
+            const onClick = () => {
+              onHide()
+              child.props.onClick?.()
+            }
+            return React.cloneElement(child, {
+              onClick,
+            })
+          } else {
+            return child
+          }
+        })
```
**question**: Maybe it would be better if we **only** allow passing `ContextualMenuItem` as children? Otherwise, if we pass another component, we will not be able to close the dropdown menu (see attached video)

### Demo
https://user-images.githubusercontent.com/3513494/220097110-75582839-bf27-4aa9-80ea-ad2df878c9a9.mp4

### Code
```javascript
<>
  {isDeleteModalOpen && (
    <DeleteSavesModal
      collection={collection}
      setIsDeleteModalOpen={setIsDeleteModalOpen}
    />
  )}

  <ContextualMenu>
    <Clickable bg="red">Hello</Clickable>

    <ContextualMenuItem onClick={handleEditList}>
      {t("collectorSaves.contextualMenu.edit")}
    </ContextualMenuItem>

    <ContextualMenuDivider />

    <ContextualMenuItem onClick={handleDeleteList}>
      {t("collectorSaves.contextualMenu.delete")}
    </ContextualMenuItem>
  </ContextualMenu>
</>
```

dimatretyak reviewed (approved):
LGTM 👍

dimatretyak commented:
@anandaroop can we fix this "jumping" state when the dropdown menu opens?

https://user-images.githubusercontent.com/3513494/220097904-326cbba5-f99f-4b0e-ae1c-c33ed2b29a1a.mp4

dimatretyak commented on src/Apps/CollectorProfile/Routes/Saves2/Components/Actions/DeleteSavesModal.tsx:
```diff
@@ -0,0 +1,105 @@
+
+          <Button
+            mb={[1, 0]}
+            onClick={async () => {
+              try {
+                await submitMutation({
+                  variables: {
+                    input: {
+                      id: collection.internalID,
+                    },
+                  },
+                  rejectIf: res => {
+                    const result = res.deleteCollection?.responseOrError
+
+                    return result?.__typename === "DeleteCollectionFailure"
+                      ? result.mutationError
+                      : false
+                  },
+                })
+
+                sendToast({
+                  variant: "success",
+                  message: t("collectorSaves.deleteListModal.success"),
+                })
+              } catch (err) {
+                console.error(err)
+                sendToast({
+                  variant: "error",
+                  message: err.message ?? "Something went wrong",
+                })
+              } finally {
+                setIsDeleteModalOpen(false)
+              }
+            }}
```
**suggestion**: move this logic to a separate function

```javascript
const handleDeletePress = async () => {
  try {
    await submitMutation({
      variables: {
        input: {
          id: collection.internalID,
        },
      },
      rejectIf: res => {
        const result = res.deleteCollection?.responseOrError

        return result?.__typename === "DeleteCollectionFailure"
          ? result.mutationError
          : false
      },
    })

    sendToast({
      variant: "success",
      message: t("collectorSaves.deleteListModal.success"),
    })
  } catch (err) {
    console.error(err)
    sendToast({
      variant: "error",
      message: err.message ?? "Something went wrong",
    })
  } finally {
    setIsDeleteModalOpen(false)
  }
}

<Button
  mb={[1, 0]}
  onClick={handleDeletePress}
>
  {t("collectorSaves.deleteListModal.delete")}
</Button>
```

anandaroop commented:
Addressed most of the feedback though I am unsure how to fix [this jumping issue](https://github.com/artsy/force/pull/11943#issuecomment-1436840027).

That ⋯ trigger is inside a `<button>` component not an `<a>`, so I'm surprised to see that jumping in the first place, unless it has something to do with the Pallete `<Dropdown>`'s  [auto-positioning behavior](https://github.com/artsy/palette/blob/e7d12034e70880369aabdbd993b3678207374eb1/packages/palette/src/elements/Dropdown/Dropdown.tsx#L106-L114)?

anandaroop reviewed (commented):
Addressed most of the feedback though I am unsure how to fix [this jumping issue](https://github.com/artsy/force/pull/11943#issuecomment-1436840027).

That ⋯ trigger is inside a `<button>` component not an `<a>`, so I'm surprised to see that jumping in the first place, unless it has something to do with the Pallete `<Dropdown>`'s  [auto-positioning behavior](https://github.com/artsy/palette/blob/e7d12034e70880369aabdbd993b3678207374eb1/packages/palette/src/elements/Dropdown/Dropdown.tsx#L106-L114)?
